p2p/pex/addrbook_test.go

```diff
@@ -417,3 +432,198 @@ func TestPrivatePeers(t *testing.T) {
 		assert.True(t, ok)
 	}
 }
+
+func testAddrBookAddressSelection(t *testing.T, bookSize int) {
+	// generate all combinations of old (m) and new addresses
+	for nOld := 0; nOld <= bookSize; nOld++ {
+		nNew := bookSize - nOld
+		dbgStr := fmt.Sprintf("book of size %d (new %d, old %d)", bookSize, nNew, nOld)
+
+		// create book and get selection
+		book, fname := createAddrBookWithMOldAndNNewAddrs(t, nOld, nNew)
+		defer deleteTempFile(fname)
+		addrs := book.GetSelectionWithBias(biasToSelectNewPeers)
+		assert.NotNil(t, addrs, "%s - expected a non-nil selection", dbgStr)
+
+		nAddrs := len(addrs)
+
+		assert.NotZero(t, nAddrs, "%s - expected at least one address in selection", dbgStr)
+
+		// find the number of old and new addresses and verify that the selection
+		// is fully populated (i.e. there are no nil addresses)
+		nOld, nNew := countOldAndNewAddrsInSelection(addrs, book)
+		assertMOldAndNNewAddrsInSelection(t, nOld, nNew, addrs, book)
+
+		// Given:
+		// n - num new addrs, m - num old addrs
+		// k - num new addrs expected in the beginning (based on bias %)
+		// i=min(n, k), aka expFirstNew
+		// j=min(m, r-i), aka expOld
+		//
+		// We expect this layout:
+		// indices:      0...i-1   i...i+j-1    i+j...r
+		// addresses:    N0..Ni-1  O0..Oj-1     Ni...
+		//
+		// There is at least one partition and at most three.
+		var (
+			k           = percentageOfNum(biasToSelectNewPeers, nAddrs)
+			expFirstNew = cmn.MinInt(nNew, k)
+			expOld      = cmn.MinInt(nOld, nAddrs-expFirstNew)
+			expNew      = nAddrs - expOld
+			expLastNew  = expNew - expFirstNew
+		)
+
+		// Verify that the number of old and new addresses are as expected
+		// XXX: Why verify expected numbers if assertMOldAndNNewAddrsInSelection
+		// which does exact comparison passes?
+		if nNew < expNew || nNew > expNew {
```

Comment:
Why verify expected numbers if assertMOldAndNNewAddrsInSelection which does exact comparison (nOld==nOld, nNew==nNew) passes?

Reply:
I think there is a problem with the assert in assertMOldAndNNewAddrsInSelection(), it was a different assert before.
If you change TestAddrBookGetSelectionWithOneNotMarkedGood() to have 90 old and 10 new the test will fail.
It should be like before, something like:
```

func assertOnlyOldAndNewAddrsInSelection(t *testing.T, addrs []*p2p.NetAddress, book *addrBook) {
	nOld, nNew := countOldAndNewAddrsInSelection(addrs, book)
	assert.Equal(t, len(addrs), nNew+nOld, "nill addresses in selection")
}
```
or maybe assertZeroNilAddrsInSelection()

Reply:
I was expecting it to fail, but it did not. Why don't we test 90 old and 10 new addresses then?

Reply:
There should be no failure. The selection may be smaller than the book. So we cannot assert that nOld in book is same as nOld in selection. same for new. The original intent there was just to detect that the selection does not include nil addresses.

Reply:
> The original intent there was just to detect that the selection does not include nil addresses.

Oh, ok then